src/matrix.cpp

```diff
@@ -0,0 +1,600 @@
+that is a matrix whose kind is any of:
+
+* :any:`MatrixKind.MaxPlusTrunc`
+* :any:`MatrixKind.MinPlusTrunc`
+* :any:`MatrixKind.NTP`
+
+:param x: the matrix.
+:type x: Matrix
+
+:returns: The threshold of *x*.
+:rtype: int
+)pbdoc");
+    }
+
+    template <typename Mat>
+    auto bind_ntp_matrix(py::module& m) {
+      using semiring_type = typename Mat::semiring_type;
+      using scalar_type   = typename Mat::scalar_type;
+      auto thing          = bind_matrix_common<Mat>(m);
+
+      thing.def(
+          py::init([](size_t                                       threshold,
+                      size_t                                       period,
+                      std::vector<std::vector<scalar_type>> const& entries) {
+            return to_matrix<Mat>(semiring<semiring_type>(threshold, period),
+                                  entries);
+          }));
+      thing.def(
+          py::init([](size_t threshold, size_t period, size_t r, size_t c) {
+            return Mat(semiring<semiring_type>(threshold, period), r, c);
+          }));
+      thing.def("one", [](Mat const& self, size_t n) {
+        return Mat::one(semiring<semiring_type>(matrix::threshold(self),
+                                                matrix::period(self)),
+                        n);
+      });
+      thing.def("one", [](Mat const& self) { return self.one(); });
+
+      m.def(
+          "period",
+          [](Mat const& x) { return matrix::period(x); },
+          py::arg("x"),
+          R"pbdoc(
+:sig=(x:Matrix)->int:
+Returns the period of an ntp matrix. This function returns the period of
+the ntp matrix *x* using its underlying semiring.
+
+:param x: the ntp matrix.
+:type x: Matrix
+
+:returns: The period of the matrix *x*.
+:rtype: int
+)pbdoc");
+      m.def(
+          "threshold",
+          [](Mat const& x) { return matrix::threshold(x); },
+          py::arg("x"),
+          R"pbdoc(
+:sig=(x:Matrix)->int:
+:only-document-once:
+Returns the threshold of a matrix over a truncated semiring.
+
+This function returns the threshold of a matrix over a truncated semiring,
+that is a matrix whose kind is any of:
+
+* :any:`MatrixKind.MaxPlusTrunc`
+* :any:`MatrixKind.MinPlusTrunc`
+* :any:`MatrixKind.NTP`
+
+:param x: the matrix.
+:type x: Mat
+
+:returns: The threshold of *x*.
+:rtype: int
+)pbdoc");
+    }
+  }  // namespace
+
+  void init_matrix(py::module& m) {
+    bind_matrix_no_semiring<BMat<>>(m);
+    bind_matrix_no_semiring<IntMat<0, 0, int64_t>>(m);
+    bind_matrix_no_semiring<MaxPlusMat<0, 0, int64_t>>(m);
+    bind_matrix_no_semiring<MinPlusMat<0, 0, int64_t>>(m);
+    bind_matrix_no_semiring<ProjMaxPlusMat<0, 0, int64_t>>(m);
+    bind_matrix_trunc_semiring<MaxPlusTruncMat<0, 0, 0, int64_t>>(m);
+    bind_matrix_trunc_semiring<MinPlusTruncMat<0, 0, 0, int64_t>>(m);
+    bind_ntp_matrix<NTPMat<0, 0, 0, 0, int64_t>>(m);
+
+    m.def(
+        "matrix_row_space_size",
```

Comment:
There is still an issue with the documentation of this where ``matrix_row_space_size(x: BMat) -> int`` appears incorrectly, but I think the fix of that will be in a separate PR.